src/jupyterlabcontroller/services/source/gar.py

```diff
@@ -0,0 +1,201 @@
+"""Image source using a Google Artifact Registry."""
+
+from __future__ import annotations
+
+from typing import Mapping
+
+from structlog.stdlib import BoundLogger
+
+from ...exceptions import InvalidDockerReferenceError, UnknownDockerImageError
+from ...models.domain.docker import DockerReference
+from ...models.domain.form import MenuImage
+from ...models.domain.kubernetes import KubernetesNodeImage
+from ...models.domain.rspimage import RSPImage, RSPImageCollection
+from ...models.v1.prepuller import PrepulledImage
+from ...models.v1.prepuller_config import GARSourceConfig, PrepullerConfig
+from ...storage.gar import GARStorageClient
+from .base import ImageSource
+
+__all__ = ["GARImageSource"]
+
+
+class GARImageSource(ImageSource):
+    """Image source using a Google Artifact Registry.
+
+    Google Artifact Registry has a much nicer API that returns all available
+    images with their aliases and hashes in one request. This means we don't
+    have to track as complex of data.
```

Comment:
This sentence missing words.  Or at least "as complex of data" doesn't read very well.